### PR TITLE
[BasicAA] Track nuw through decomposed expressions

### PR DESCRIPTION
**Pull request by nikic**
When we decompose the GEP offset expression, and the arithmetic is not performed using nuw operations, we cannot retain the nuw flag on the decomposed GEP.

For example, if we have `gep nuw p, (a-1)`, this is not at all the same as `gep nuw (gep nuw p, a), -1`.

Fix this by tracking NUW through linear expression decomposition, similarly to what we already do for the NSW flag.

This fixes the miscompilation reported in https://github.com/llvm/llvm-project/pull/105496#issuecomment-2315322220.

### CODE REVIEW

File: llvm/lib/Analysis/BasicAliasAnalysis.cpp [hunk 1]
```diff
@@ -375,24 +375,28 @@ struct LinearExpression {
...
     Offset = APInt(BitWidth, 0);
   }
 
-  LinearExpression mul(const APInt &Other, bool MulIsNSW) const {
+  LinearExpression mul(const APInt &Other, bool MulIsNUW, bool MulIsNSW) const {
     // The check for zero offset is necessary, because generally
     // (X +nsw Y) *nsw Z does not imply (X *nsw Z) +nsw (Y *nsw Z).
     bool NSW = IsNSW && (Other.isOne() || (MulIsNSW && Offset.isZero()));
-    return LinearExpression(Val, Scale * Other, Offset * Other, NSW);
+    bool NUW = IsNUW && (Other.isOne() || MulIsNUW);
```
Comment (nikic) [inline llvm/lib/Analysis/BasicAliasAnalysis.cpp:398]: Proof that mul nuw always distributes (unlike mul nsw): https://alive2.llvm.org/ce/z/MZd_JJ

File: llvm/lib/Analysis/BasicAliasAnalysis.cpp [hunk 2]
```diff
@@ -444,20 +448,22 @@ static LinearExpression GetLinearExpression(
         E = GetLinearExpression(Val.withValue(BOp->getOperand(0), false), DL,
                                 Depth + 1, AC, DT);
         E.Offset += RHS;
+        E.IsNUW &= NUW;
```
Comment (nikic) [inline llvm/lib/Analysis/BasicAliasAnalysis.cpp:451]: Proof for add nuw associativity: https://alive2.llvm.org/ce/z/oiNosh

And counter-proof that the current NSW code here is incorrect: https://alive2.llvm.org/ce/z/Ga8-CT

We are missing a pre-condition: https://alive2.llvm.org/ce/z/wkP_yw

But this is unrelated to this patch...

Comment (nikic) [reply llvm/lib/Analysis/BasicAliasAnalysis.cpp:451]: I have trouble actually coming up with a test case that currently gets miscompiled due to the NSW issue ... ideas welcome :)

**Review by hazzlim — APPROVED** [review]
LGTM - thanks for getting this fixed!
